### DISCUSSION TITLE
Note on duplicate code in printf/printf_P

### DISCUSSION THREAD

**Discussion by greiman** [earlephilhower/arduino-pico#2122] [General]
This is not an issue, just information on how I handled a similar problem.

While looking at whether printf is implemented in various Arduino board packages I noticed [this](https://github.com/earlephilhower/ArduinoCore-API/blob/ece6e68f29c6f406a4434659bcbcfe558baaa3a9/api/Print.cpp#L262) comment.

> // TODO - must be better way than cut-n-paste!

I recently had the same problem for a wrapper I wrote to add printf and printf_P to classes derived from Print.

Here is my solution modified to remove cut-n-paste from your version of Print. It's one way to avoid duplicate code.

```
// Add the definition of this function to Print.h.
size_t Print::printf_common(const char *format, va_list arg) {
  char temp[64];
  char *buffer = temp;
  va_list arg_temp;
  va_copy(arg_temp, arg);
  size_t len = vsnprintf(temp, sizeof(temp), format, arg_temp);
  va_end(arg_temp);
  if (len > sizeof(temp) - 1) {
    buffer = new char[len + 1];
    if (!buffer) {
      return 0;
    }
    vsnprintf(buffer, len + 1, format, arg);
  }
  len = write((const uint8_t *)buffer, len);
  if (buffer != temp) {
    delete[] buffer;
  }
  return len;
}

size_t Print::printf(const char *format, ...) {
  va_list arg;
  va_start(arg, format);
  size_t rtn = printf_common(format, arg);
  va_end(arg);
  return rtn;
}

size_t Print::printf_P(const char *format, ...) {
  va_list arg;
  va_start(arg, format);
  size_t rtn = printf_common(format, arg);
  va_end(arg);
  return rtn;
}
```
Notice I used `va_copy` so arg is not used twice.  I was surprised to learn `va_lis`t is not a simple pointer to the stack in some implementations where arguments may be passed in registers and only one pass over the arguments will be allowed.

See this for the recommended way to use va_copy in the future. 

https://en.cppreference.com/w/cpp/utility/variadic/va_copy

**Comment by earlephilhower** [comment]
That's a better way definitely, thanks!  I tried initially with `__attribute__((alias))` but could never seem to get it working due to C++ name mangling.

**Comment by greiman** [comment]
I just found a problem with the above, printf and vsnprintf return int:

> The number of characters written if successful or negative value if an error occurred.

https://en.cppreference.com/w/c/io/vfprintf

So I am modifying my Print wrapper to check for a negative return from vsnprintf/vsnprintf_P or if write(const char*, size_t) has an error and return a negative values from my printf/printf_P.

I found this when I tried to write a template for either vsnprintf or vsnprintf_P to support AVR PSTR():

`template<size_t (*FCN)(char*, size_t, const char*, va_list)>`

I was mystified why it didn't work.  Then I finally found the return problem.
